### ISSUE TITLE
Proposal to refactor style API

### ISSUE THREAD

**Issue by antonok-edm** [maps4print/azul#67]
## Description
Currently, the only way to style azul applications is with CSS syntax that is loaded as an extra file or string.

## Describe your solution
#29 and #56 both describe proposals to load and parse custom syntax for DOM layout. There are a lot of different ways to write a hierarchical data structure, so I agree with the idea of keeping the public interface of azul minimal but open for extension -- just supplying the datatypes in the main library is a great way to support that.

Likewise, there are many different ways to specify the style of an application (Sass, SCSS, Less, Stylus, ...). Opening up an API for azul's style datatypes would make these alternative parsers a viable option, while also making for a more consistent programming interface. Keeping style parsing in separate crates could reduce the burden of support, and open the possibility of hotswapping alternate implementations.

While I don't personally have strong feelings about inline styles, this change would also make those fairly simple to expose from an API standpoint. I'd envision something like a `with_style(...)` method for DOM nodes.

Ultimately, my proposal is to expose style specifiers for azul apps in a similar way to how DOM trees are currently assembled, without needing to write any CSS. For example, something like the following would be valid code:

```rust
let rules = vec![
    Rule::new(Selector::new(StylePath::Global).with_child(Selector::new(StylePath::Div).with_class("node")), Style::Background(Color { r: 0, g: 0, b: 0, a: 255 }))
];

let app = App::new(mydatamodel, AppConfig::default());
app.run(Window::new(WindowCreateOptions::default(), AppStyle::new(rules)).unwrap()).unwrap();
```

Then, a separate crate (perhaps in the same workspace), `azul-css`, could produce that AppStyle object from CSS (in a file, string, macro, ...) like the following:

```css
* div.node {
    background-color: black;
}
```

## Are there alternatives or drawbacks?
I'd be open to suggestions on syntax/API - consider the above example a rough draft. 

In terms of drawbacks -- the base syntax would be more verbose than it is currently, but that would be dealt with by parser crates for anyone concerned.

## Is this a breaking change
I think the most "correct" way to implement this would be as a breaking change. I suppose it would be possible to implement this in a backward compatible way, but it probably wouldn't result in a clean API.

## Additional notes
Hot-reload could be managed by exposing a `HotReloadable` trait for 3rd party style suppliers to implement, which would allow azul to periodically request an `Option<AppStyle>` by callback. For example,

```rust
app.run(Window::new(WindowCreateOptions::default(), AppStyle::hot_reload(azul_css::hot_reloader("/path/to/file.css"))).unwrap()).unwrap();
```

**Comment by fschutt** [comment]
Ok, but you are aware that you can already do that? A CSS file with rules can be created programmatically from a `Vec<CssRuleBlock>` and you can construct a `CssRuleBlock` in pure Rust code. So I'm not sure what you're asking for, because this is already implemented [here](https://github.com/maps4print/azul/blob/d75f2167c7c7ef5faff5627d25976a3fa0669f70/src/css.rs#L41).

As for putting the style directly on the DOM nodes ("inline styles") - no I don't think this is a good idea:

1. The CSS is not immutable anymore, it can change every frame and has to be re-built every frame, which is a performance loss.
2. It prevents sharing styles between DOM nodes via pointers (to prevent memory bloat issues).
3. It prevents hot-reloading from a file since the CSS is now owned by the DOM nodes.
4. Styles are harder to share between nodes - it's easier to put extra classes on the DOM nodes than passing around the styling information in the layout() function all the time.

As I understand it, you want the CSS parser in a seperate crate - which is a good idea, because other projects (i.e. SVG parsers) may also need a CSS parser, but not a full rendering library. But all you'd need to do for that is to seperate out the `css.rs` file and the `css_parser.rs` file into a seperate crate.

**Comment by antonok-edm** [comment]
> are you aware that you can already do that?

Well, half of it's done already then... I used that as inspiration but I didn't realize it was already a public-facing API. I suppose I can still write some documentation about how that works, since I wasn't able to find any.

The separation of crates is part of my proposal, but I'm also imagining abstracting the supported styles away from specifically CSS. I guess for the most part things are functionally pretty well-abstracted away at this point and it's mostly just bikeshedding about whether things are named `CssRuleBlock` vs. just `RuleBlock`. The one thing that stands out to me is the presence of both `Background` and `BackgroundColor`, since the CSS background property can take a color as well.

**Comment by antonok-edm** [comment]
@ErichDonGubler Not quite sure what you're asking. I haven't looked at that crate in too much depth, but the README says

> This crate will break compatibility regularly and often as it matures

It looks more like the creator's personal project than something ready to be used in a production GUI library.

That said, the purpose of this issue is to split css *parsing* logic into a separate crate. One thing that enables is a very easy way to swap in an alternate parsing implementation. If you felt inclined to build a wrapper around the crate you found which converted its data representation into the one used for azul's styles, that would absolutely be an option.

**Comment by fschutt** [comment]
Yeah, one of the things I have to re-iterate is that *parsing* is a very small problem compared to every other issue. The thing is that it doesn't help if you have a parser, but that parser doesn't affect anything (the implementation of the parsed property is missing). So I try to only expose properties that are actually implemented, otherwise it gives a false sense of promise. There are things that azul will likely never have (such as CSS variables - because there are dynamic properties already). So it's better to say "that's an error, not supported", rather than just parsing it.

I have not seen the css crate before - I don't think it's very actively developed, but it might still be good to use some implementation details. However, as I said, parsing is less of a problem than implementing it.

@antonok-edm This probably doesn't fit into this discussion, but if you're searching for something to do, text shadows, text gradients, image filters (at least blur) and image shadows (not box shadows, real image shadows that incorporate the image alpha component) should be fairly easy to implement (since the parsers already exist and you just have to pass them to webrender).

**Comment by antonok-edm** [comment]
@fschutt Agree, and I think moving parsing to an external component will solidify that. Right now with CSS parsing built in, it's easy for users to assume azul is "broken" if it doesn't support a CSS feature. Passing the library a stricter, type-constrained datatype instead means that it should be impossible to ask azul itself to render something it doesn't know how to handle.

Things like CSS variables would be a great candidate for the kind of "post-processing" that an external parser can do. AFAIK it's impossible to do dynamic variables in CSS anyways, only static ones for the sake of deduplicating constants.

Thanks for the recommendations, I can start to look into those once I finish splitting the parser out.

**Comment by fschutt** [comment]
Closed in #70
